dvc/api/data.py

```diff
@@ -73,6 +73,7 @@ def open(  # noqa, pylint: disable=redefined-builtin
     remote: Optional[str] = None,
     mode: str = "r",
     encoding: Optional[str] = None,
+    config: Optional[Dict[str, Any]] = None,
```

Comment:
Or, we could just name it `config` but we'll be passing `{"core": {"remote": remote}, "remote": {remote: config}}.

Reply:
Agreed. I would simplify even more to the structure of a single remote's config, like `open(..., remote_config={"url": ..}`). If there is also a `remote` arg, we can merge it with the config for that remote. if not, we can merge it with the default remote.

Related: https://github.com/iterative/dvc.org/pull/4628#discussion_r1231667869.

Reply:
`config` is more powerful though and you could set new default remote there in one place as well. Sure, we could also do `remote_config` but that's more niche and we will need to tell what remote we want that to apply to. Remember that one could have multiple remotes in the repository (plus also for remote notation), which `remote_config` won't handle. I chose `config` because it is the most complete solution, while we could add edge-case params later in the future if there will be demand.

Reply:
> we will need to tell what remote we want that to apply to

That already exists, there's `remote` kwarg for that. 


> Remember that one could have multiple remotes in the repository

This is not applicable for `open/read` API though as they are about single file, right?

Reply:
> It should be simple to use too.

Totally, hence why I'm open to any alternatives like `remote_config`, but I want to give the capability right now. I don't think those contradict each other.

> That already exists, there's remote kwarg for that.

That will only change the default config, but your outputs might be configured to use custom remotes which are not set as default.

> This is not applicable for open/read API though as they are about single file, right?

That's right, but that would require us to find what remote corresponds to that particular file, which we don't do.

Reply:
> That will only change the default config, but your outputs might be configured to use custom remotes which are not set as default.

👍 That's a good point I hadn't considered.

> That will only change the default config

So it won't override a `remote:` field in `.dvc`? We should clarify that in the docs.

Maybe it is worth having both `config` and `remote_config`. Or we could make `remote` take either a string or dict.

Reply:
> So it won't override a remote: field in .dvc? We should clarify that in the docs.

It won't, but you can change the config for that remote. So the way it would work is - you need to know the name of that remote for that output (by looking at the corresponding dvc file) and then pass the configuration for it here. The mechanism is the same as using `.dvc/config.local` really.

Reply:
So to set a completely new remote, you would need something like:

```python
with dvc.api.open(
    "data",
    remote="newremote",
    config={
        "remote": {
            "newremote": {
                "url": ...,
                ...
            }
        }
    }
) as f:
```

Can we add `remote_config` so it works like this?

```python
with dvc.api.open(
    "data",
    remote="newremote",
    remote_config={
        "url": ...,
        ...
    }
) as f:
```

`remote` sets the default remote, and `remote_config` merges with the config for the default remote. If you need control of additional remotes, you can use `config`.

Reply:
@dberenbaum I would rather just:

```
with dvc.api.open(
    "data",
    remote="newremote",
    config={
        "core": {"remote": "newremote"},
        "remote": {
            "newremote": {
                "url": ...,
                ...
            }
        }
    }
) as f:
```

because that's how you would do it in your .dvc/config[.local]

Reply:
@efiop What is the point of `remote="newremote"` there? 

I agree with @skshetry that it requires too much knowledge of how the config is structured for simple use cases and is too verbose. `config` makes sense to have, but I don't think it's the best way to do something like add credentials to your remote.

Reply:
> @efiop What is the point of remote="newremote" there?

That was a mistake, I removed that in the edit. Please take another look.

Reply:
I think that looks easier for you because you are familiar with the dvc remote structure. I doubt that's more obvious to most users. `config` is more flexible and explicit, but in many cases users probably just need something like `remote_config={"token": ...}` and setting up that whole section above feels like overkill.

Reply:
> My logic with `remote` is not that I'm familair with remote structure but rather that users are familiar with dvc config, since that's the way they would be changing stuff manually.

Yup, sorry, that's what I meant. I doubt they know the config structure without looking it up.

`config` could be useful for more advanced scenarios, like setting options for multiple remotes. Only the `remote` section of it makes sense, though, right? I don't know that I see enough need to expose it publicly. I would rather have `remote_config`, but I'm not strictly against keeping both.

Reply:
Having both `remote` and `remote_config` seems a bit ugly too. Ideally only `remote_config` would be enough (so you don't have to remember or make up remote names), but that requires some internal additions, so maybe in the future. Adding `remote_config` alongside `remote` will also require some time, so I would rather indeed keep `config` for now as the most powerful option that is great to have around even in the future.

One thing for sure - it is probably time to document config `dict` format in addition to the regular file config format, this will be useful for everyone long term.

Reply:
> Ideally only `remote_config` would be enough

Agreed. I think it can be enough in most cases where you want to use the default remote. Take this example:

```python
with dvc.api.open("data", remote_config={"token": ...}) as f:
```

This is enough to set additional config options for the default remote, which I think is the most likely use case. How hard is it to add?

Compare that to how it looks with `config`:

```python
with dvc.api.open("data", config={"remote": {"myremote": {"token": ...}}}) as f:
```

`config` is not only longer, but users have to know the config structure and the name of the default remote.

> I would rather indeed keep `config` for now as the most powerful option that is great to have around even in the future.

Is there any way it will be used for anything besides remote config? I can't see any other config section that would make sense to override.

Reply:
> How hard is it to add?

For default - not hard, but currently if something is using non-default one it will work, and supporting that is the time-consuming part. I'm totally on the same page with you, `remote_config` is useful, just not sure it is worth investing into right now since `config` is already there.

> Is there any way it will be used for anything besides remote config? I can't see any other config section that would make sense to override.

Remote config is the prime use case, but `config` is the most powerful mechanism that can allow one to get out of very sticky situations (e.g. if one drops git from the repo but still wants to use it he can set `no_scm` through this, this seems useful during deployment somewhere).

Reply:
What's the level of effort to also add `remote_config`?

Reply:
Let's discuss when we talk tomorrow

Reply:
> config is the most powerful mechanism that can allow one to get out of very sticky situations

@efiop, I think you are looking at it from a maintainer's point of view rather than the user. We can provide powerful mechanism in `DvcFileSystem`, `read/open` should be simpler. `remote_config=` solves most of the problem imo.

Creating a new remote, and/or providing a way to configure `dvc` repo is more of a niche usecase. Also, I don't feel comfortable exposing the whole config schema to the users. It has a large surface area, and can have unintended effect when not use correctly.